iroha_network/src/mock.rs

```diff
@@ -0,0 +1,355 @@
+        mut handler: H,
+    ) -> Result<(), String>
+    where
+        H: FnMut(State<S>, Request) -> F,
+        F: Future<Output = Result<Response, String>>,
+    {
+        let mut buffer = [0u8; BUFFER_SIZE];
+        let read_size = stream
+            .read(&mut buffer)
+            .await
+            .expect("Request read failed.");
+        let bytes: Vec<u8> = buffer[..read_size].to_vec();
+        let request: Request = bytes
+            .try_into()
+            .map_err(|e: Box<dyn Error>| e.to_string())?;
+        let response: Vec<u8> = handler(state, request).await?.into();
+        stream
+            .write_all(&response)
+            .await
+            .map_err(|e| e.to_string())?;
+        stream.flush().await.map_err(|e| e.to_string())?;
+        Ok(())
+    }
+}
+
+#[derive(Eq, PartialEq, Debug, Clone)]
+pub struct Request {
+    uri_path: String,
+    payload: Vec<u8>,
+}
+
+impl Request {
+    /// Creates a new request
+    ///
+    /// # Arguments
+    ///
+    /// * uri_path - corresponds to [URI syntax](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier)
+    /// `path` part (e.g. "/commands")
+    /// * payload - the message in bytes
+    ///
+    /// # Examples
+    /// ```
+    /// use iroha_network::prelude::*;
+    ///
+    /// let request = Request::new("/commands".to_string(), "some_message".to_string().into_bytes());
+    /// ```
+    pub fn new(uri_path: String, payload: Vec<u8>) -> Request {
+        Request { uri_path, payload }
+    }
+
+    pub fn url(&self) -> &str {
+        &self.uri_path[..]
+    }
+
+    pub fn payload(&self) -> &[u8] {
+        &self.payload[..]
+    }
+}
+
+impl From<Request> for Vec<u8> {
+    #[log]
+    fn from(request: Request) -> Self {
+        let mut bytes = Vec::new();
+        bytes.extend_from_slice(request.uri_path.as_bytes());
+        bytes.extend_from_slice(b"\r\n");
+        bytes.extend(request.payload.into_iter());
+        bytes
+    }
+}
+
+impl TryFrom<Vec<u8>> for Request {
+    type Error = Box<dyn Error>;
+
+    #[log]
+    fn try_from(mut bytes: Vec<u8>) -> Result<Request, Box<dyn Error>> {
+        let n = bytes
+            .iter()
+            .position(|byte| *byte == b"\n"[0])
+            .expect("Request should contain \\r\\n sequence.")
+            + 1;
+        let payload: Vec<u8> = bytes.drain(n..).collect();
+        Ok(Request {
+            uri_path: String::from_utf8(bytes[..(bytes.len() - 2)].to_vec())?,
+            payload,
+        })
+    }
+}
+
+#[derive(Debug, PartialEq, Io, Encode, Decode)]
+pub enum Response {
+    Ok(Vec<u8>),
+    InternalError,
+}
+
+impl Response {
+    pub fn empty_ok() -> Self {
+        Response::Ok(Vec::new())
+    }
+}
+
+pub mod prelude {
+    //! Re-exports important traits and types. Meant to be glob imported when using `iroha_network`.
+
+    #[doc(inline)]
+    pub use crate::{AsyncStream, Network, Request, Response, State};
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Maybe we can can combine the tests in the main and mock modules, and just conditionally compile the import? The tests there duplicate each other.

Reply:
They should duplicate each other to check that both modules work identical. 
Conditional compilation looks like additional build step - I think it's okay to compile both modules during tests, having only one compiled in client app.

If you have an elegant proposal - please describe it in more details.

Reply:
If we have identical tests in two places, later we may edit one of them and forget about another one. But yes, I don't have a workaround for another build step right now...

Reply:
Moreover - we will need to update some methods code also( When we will work with network more - it should be rewritten a bit